---
title: "R Notebook"
output:
  html_document:
    df_print: paged
---


```{r global_options}
library(knitr)
knitr::opts_chunk$set(fig.path='Figs/')
```


```{r}

#install.packages("devtools")
library(devtools)
library(maftools)
```
# Reading MAF files.
Reading MAF files and storing in a MAF object.
```{r}
  #path to TCGA LAML MAF file
laml.maf = system.file('extdata', 'tcga_laml.maf.gz', package = 'maftools') 
#clinical information containing survival information and histology. This is optional
laml.clin = system.file('extdata', 'tcga_laml_annot.tsv', package = 'maftools') 

laml = read.maf(maf = laml.maf, clinicalData = laml.clin)
```
# MAF object
Summarised data of the MAF object
```{r}
#Typing laml shows basic summary of MAF file.
laml
#Shows sample summary
getSampleSummary(laml)
#Shows gene summary.
getGeneSummary(laml)
#shows clinical data associated with samples
getClinicalData(laml)
#Shows all fields in MAF
getFields(laml)
#Writes maf summary to an output file with basename laml.
write.mafSummary(maf = laml, basename = 'laml')
```
# Visualization
# 7.1 Plotting MAF summary
Using plotmafSummary to plot the summary of the maf file, which displays number of variants in each sample as a stacked barplot and variant types as a boxplot summarized by Variant_Classification.
```{r}
plotmafSummary(maf = laml, rmOutlier = TRUE, addStat = 'median', dashboard = TRUE, titvRaw = FALSE)

```
# 7.2 Oncoplots
# 7.2.1 Drawing oncoplots
Oncoplots show better represention of maf object
```{r}
#oncoplot for top ten mutated genes.
oncoplot(maf = laml, top = 10)
```
# 7.3 Oncostrip
oncostrip functiond raws mutations in each sample can be used to draw any number of genes using top or genes arguments.
```{r}
oncostrip(maf = laml, genes = c('DNMT3A','NPM1', 'RUNX1'))
```
# 7.4 Transition and Transversions
titv function classifies SNPs into Transitions and Transversions and returns  that can also be visualized as a boxplot showing overall distribution of six different conversions and as a stacked barplot showing fraction of conversions in each sample.
```{r}
laml.titv = titv(maf = laml, plot = FALSE, useSyn = TRUE)
#plot titv summary
plotTiTv(res = laml.titv)
```
# 7.5 Lollipop plots for amino acid changes
Lollipop plots are simple and most effective way showing mutation spots on protein structure. Many oncogenes have a preferential sites which are mutated more often than any other locus. We can draw such plots using the function lollipopPlot. LollipopPlot function is used to plot this plot with the amino acid changes information in the maf file. 
```{r}
#lollipop plot for DNMT3A, which is one of the most frequent mutated gene in Leukemia.
lollipopPlot(maf = laml, gene = 'DNMT3A', AACol = 'Protein_Change', showMutationRate = TRUE)
```
# 7.5.1 Labelling points.
Labelling the plot with Labelpos
```{r}
lollipopPlot(maf = laml, gene = 'KIT', AACol = 'Protein_Change', labelPos = 816, refSeqID = 'NM_000222')
```
# 7.6 Rainfall plots
Hyper mutated genomic regions can be visualized by plotting inter variant distance on a linear genomic scale. These plots are called rainfall plots and are drawn with plots using rainfallPlot
```{r}
brca <- system.file("extdata", "brca.maf.gz", package = "maftools")
brca = read.maf(maf = brca, verbose = FALSE)

rainfallPlot(maf = brca, detectChangePoints = TRUE, pointSize = 0.6)
```
# 7.7 Compare mutation load against TCGA cohorts
It is informative to see how mutation load in given maf stands against TCGA cohorts.This is obtained with function tcgaComapre which draws distribution of variants compiled from over 10,000 WXS samples across 33 TCGA landmark cohorts.
```{r}
laml.mutload = tcgaCompare(maf = laml, cohortName = 'Example-LAML')

```
# 7.8 Plotting VAF
```{r}
plotVaf(maf = laml, vafCol = 'i_TumorVAF_WU')
```
# 7.9 Genecloud
 word cloud plot for mutated genes with the function geneCloud. Size of each gene is proportional to the total number of samples in which it is mutated/altered.
```{r}
geneCloud(input = laml, minMut = 3)

```
# 9 Analysis
# 9.1 Somatic Interactions
Mutually exclusive or co-occurring set of genes can be detected using somaticInteractions function, which performs pair-wise Fisher’s Exact test to detect such significant pair of genes
```{r}
#exclusive/co-occurance event analysis on top 10 mutated genes. 
somaticInteractions(maf = laml, top = 25, pvalue = c(0.05, 0.1))

```
# 9.2 Detecting cancer driver genes based on positional clustering
Concept of detecting the cancer driven gene based on positional clustring is based on the fact that most of the variants in cancer causing genes are enriched at few specific loci (aka hot-spots). This method takes advantage of such positions to identify cancer genes

```{r}
laml.sig = oncodrive(maf = laml, AACol = 'Protein_Change', minMut = 5, pvalMethod = 'zscore')
head(laml.sig)



plotOncodrive(res = laml.sig, fdrCutOff = 0.1, useFraction = TRUE)
```
#  9.3 Adding and summarizing pfam domains
pfamDomains adds pfam domain information to the amino acid changes and also summarizes amino acid changes according to the domains that are affected. 
```{r}
laml.pfam = pfamDomains(maf = laml, AACol = 'Protein_Change', top = 10)
#Protein summary (Printing first 7 columns for display convenience)
laml.pfam$proteinSummary[,1:7, with = FALSE]
#Domain summary (Printing first 3 columns for display convenience)
laml.pfam$domainSummary[,1:3, with = FALSE]
```
# 9.4 Pan-Cancer comparison
comparing mutSig results against this pan-can list of significantly mutated genes to see genes specifically mutated in given cohort. This function requires MutSigCV results 
```{r}
#MutsigCV results for TCGA-AML
laml.mutsig <- system.file("extdata", "LAML_sig_genes.txt.gz", package = "maftools")
pancanComparison(mutsigResults = laml.mutsig, qval = 0.1, cohortName = 'LAML', inputSampleSize = 200, label = 1)
```
#9.5 Survival analysis
# 9.5.1 Mutation in any given genes
Function mafSurvive performs survival analysis and draws kaplan meier curve by grouping samples based on mutation status of user defined gene(s) or manually provided samples those make up a group. This function requires input data to contain Tumor_Sample_Barcode (make sure they match to those in MAF file), binary event (1/0) and time to event.
```{r}
#Survival analysis based on grouping of DNMT3A mutation status
mafSurvival(maf = laml, genes = 'DNMT3A', time = 'days_to_last_followup', Status = 'Overall_Survival_Status', isTCGA = TRUE)
```
# 9.5.2 Predict genesets associated with survival
```{r}
#Using top 20 mutated genes to identify a set of genes (of size 2) to predict poor prognostic groups
prog_geneset = survGroup(maf = laml, top = 20, geneSetSize = 2, time = "days_to_last_followup", Status = "Overall_Survival_Status", verbose = FALSE)
print(prog_geneset)
mafSurvGroup(maf = laml, geneSet = c("DNMT3A", "FLT3"), time = "days_to_last_followup", Status = "Overall_Survival_Status")
```
# 9.6 Comparing two cohorts (MAFs)
comparing two different cohorts to detect such differentially mutated genes.difference between two cohorts (in this case primary and relapse APL) can be detected using function mafComapre, which performs fisher test on all genes between two cohorts to detect differentially mutated genes.
```{r}
#Primary APL MAF
primary.apl = system.file("extdata", "APL_primary.maf.gz", package = "maftools")
primary.apl = read.maf(maf = primary.apl)
#Relapse APL MAF
relapse.apl = system.file("extdata", "APL_relapse.maf.gz", package = "maftools")
relapse.apl = read.maf(maf = relapse.apl)

#Considering only genes which are mutated in at-least in 5 samples in one of the cohort to avoid bias due to genes mutated in single sample.
pt.vs.rt <- mafCompare(m1 = primary.apl, m2 = relapse.apl, m1Name = 'Primary', m2Name = 'Relapse', minMut = 5)
print(pt.vs.rt)

```
# 9.6.1 Forest plots
```{r, error=TRUE}
forestPlot(mafCompareRes = pt.vs.rt, pVal = 0.1, color = c('royalblue', 'maroon'), geneFontSize = 0.8)

```

# 9.6.2 Co-onco plots
 coOncoplot function takes two maf objects and plots them side by side for better comparison.
```{r}
genes = c("PML", "RARA", "RUNX1", "ARID1B", "FLT3")
coOncoplot(m1 = primary.apl, m2 = relapse.apl, m1Name = 'PrimaryAPL', m2Name = 'RelapseAPL', genes = genes, removeNonMutated = TRUE)
```

# 9.6.3 Lollipop plot-2
```{r}
lollipopPlot2(m1 = primary.apl, m2 = relapse.apl, gene = "PML", AACol1 = "amino_acid_change", AACol2 = "amino_acid_change", m1_name = "Primary", m2_name = "Relapse")
```
# 9.7 Clinical enrichment analysis
Clinical Enrichment performs various groupwise and pairwise comparisions to identify enriched mutations for every category within a clincila feature. 
```{r}
fab.ce = clinicalEnrichment(maf = laml, clinicalFeature = 'FAB_classification')

#Results are returned as a list. Significant associations p-value < 0.05
fab.ce$groupwise_comparision[p_value < 0.05]
plotEnrichmentResults(enrich_res = fab.ce, pVal = 0.05)
```
# 9.8 Drug-Gene Interactions
drugInteractions function checks for drug–gene interactions and gene druggability information compiled from Drug Gene Interaction database.
```{r}
dgi = drugInteractions(maf = laml, fontSize = 0.75)
dnmt3a.dgi = drugInteractions(genes = "DNMT3A", drugs = TRUE)
#Printing selected columns.
dnmt3a.dgi[,.(Gene, interaction_types, drug_name, drug_claim_name)]
```
# 9.9 Oncogenic Signaling Pathways
```{r}
OncogenicPathways(maf = laml)
PlotOncogenicPathways(maf = laml, pathways = "RTK-RAS")
```
# 9.11 Mutational Signatures
signatures can be extracted by decomposing matrix of nucleotide substitutions, classified into 96 substitution classes based on immediate bases surrounding the mutated base. 
```{r}
# BSgenome.Hsapiens.UCSC.hg19 package
#if (!requireNamespace("BiocManager", quietly = TRUE))
 #   install.packages("BiocManager")
#BiocManager::install("BSgenome.Hsapiens.UCSC.hg19")

# NMF and pheatmap packages
#install.packages(c("pheatmap", "NMF"))
```

```{r}
library(BSgenome.Hsapiens.UCSC.hg19, quietly = TRUE)
laml.tnm = trinucleotideMatrix(maf = laml, prefix = 'chr', add = TRUE, ref_genome = "BSgenome.Hsapiens.UCSC.hg19")

```
# 9.11.2 Differences between APOBEC enriched and non-enriched samples
plotApobecDiff function takes APOBEC enrichment scores estimated by trinucleotideMatrix and classifies samples into APOBEC enriched and non-APOBEC enriched. it compares these two groups to identify differentially altered genes
```{r}
plotApobecDiff(tnm = laml.tnm, maf = laml, pVal = 0.2)

```
# 9.11.3 Signature analysis
estimateSignatures runs NMF on a range of values and measures the goodness of fit - in terms of Cophenetic correlation.
```{r}
library('NMF')
laml.sign = estimateSignatures(mat = laml.tnm, nTry = 6, pConstant = 0.1)
laml.sig = extractSignatures(mat = laml.tnm, n = 3, pConstant = 0.1)
plotCophenetic(res = laml.sign)
laml.sig = extractSignatures(mat = laml.tnm, n = 3,pConstant = 0.1)
#Compate against original 30 signatures 
laml.og30.cosm = compareSignatures(nmfRes = laml.sig, sig_db = "legacy")
#Compate against updated version3 60 signatures 
laml.v3.cosm = compareSignatures(nmfRes = laml.sig, sig_db = "SBS")
library('pheatmap')
pheatmap::pheatmap(mat = laml.og30.cosm$cosine_similarities, cluster_rows = FALSE, main = "cosine similarity against validated signatures")
maftools::plotSignatures(nmfRes = laml.sig, title_size = 0.8)
```
# 9.11.4 Signature enrichment analysis
signatureEnrichment funtion identifies mutations enriched in every signature identified
```{r}
laml.se = signatureEnrichment(maf = laml, sig_res = laml.sig)
plotEnrichmentResults(enrich_res = laml.se, pVal = 0.05)

```

